utils/utf8.hh

```diff
@@ -64,6 +65,54 @@ inline std::optional<size_t> validate_with_error_position(bytes_view string) {
     return validate_with_error_position(data, len);	
 }
 
+std::optional<size_t> validate_with_error_position_fragmented(const FragmentRange auto& fr) {
+    uint8_t partial_codepoint[4];
+    size_t partial_filled = 0;
+    size_t partial_more_needed = 0;
+    size_t bytes_validated = 0;
+    for (auto&& frag : fr) {
+        auto data = reinterpret_cast<const uint8_t*>(frag.data());
+        auto len = frag.size();
+        if (partial_more_needed) {
+            // Tiny loop (often zero iterations), don't call memcpy
+            while (partial_more_needed && len) {
+                partial_codepoint[partial_filled++] = *data++;
+                --partial_more_needed;
+                --len;
+            }
+            if (!partial_more_needed) {
+                // We accumulated a codepoint that straddled two oe more fragments,
+                // validate it now.
+                auto pvr = internal::validate_partial(partial_codepoint, partial_filled);
+                if (pvr.error) {
+                    return {bytes_validated};
+                }
+                bytes_validated += partial_filled;
+                partial_filled = partial_more_needed = 0;
+            }
+            if (!len) {
```

Comment:
Super nit: it is guaranteed that after a while loop above either partition_more_needed == 0 or len == 0 so here you can just do else and inside this else do assert (!len).

Reply:
Ok actually no. It is also possible for both partial_more_needed and len to be zero so `else` wouldn't work.

Reply:
Correct.